cfpb/capital-framework
# Add Yarn and Lerna to project, remove CI

Adds Lerna for monorepo management and Yarn for it's neato workspaces and caching features.

Check out the [contributing docs](https://github.com/cfpb/capital-framework/blob/0e95345db34ad8e9a6349af8468ca4101b311292/CONTRIBUTING.md#changing-the-codebase) to see how these changes affect CF development. Spoiler alert: The only thing you'll have to do differently is remember to use `yarn` instead of `npm`.

### Notable Changes

1. The `src/` directory in the project's root has been renamed to `packages/`.
1. All the Travis CI stuff has been removed. We'll manually publish components (which is [easy with Lerna](https://github.com/cfpb/capital-framework/blob/7f1e16add97dd5abd91f6d9b72feba4ac149b057/CONTRIBUTING.md#release-management)) for the time being and we can move the task back into Travis if/when we're ready.
1. No more npm or `package-lock.json`. Node modules are cached in the git repo.
1. No more scary scripts for publishing/linking. Yarn and Lerna do most of it natively.

### Testing

1. Pull down this branch (might take a minute with all the cached packages).
1. `rm -rf node_modules/` to ensure you're starting with a clean slate.
1. `yarn` to install (no more `npm install`)
1. `yarn test` to confirm the tests pass as normal.
1. `yarn run cf-link` to link all the components
1. Edit `/packages/cf-core/src/cf-vars.less` and set the text color on line 13 to something silly like `red`.
1. `cd ~/wherever/cfgov-refresh`
1. `yarn link cf-core`
1. `gulp clean && gulp build`
1. Start up cfgov-refresh and open http://0.0.0.0:8000/. Text should now be red, indicating that it's correctly picking up your local version of `cf-core`.

### Unanswered questions

1. It's currently configured to use [fixed](https://github.com/lerna/lerna#how-it-works) versioning. Should we use independent versioning instead? We've been split on this. Let's keep discussing. **Update:** We discussed this and are happy with fixed versioning.
